File: C++/2-Medium/00236._二叉树的最近公共祖先/00236._二叉树的最近公共祖先.cpp
#include <iostream>
#include <vector>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    // 寻找最近公共祖先的递归函数
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root || root == p || root == q) return root; // 基础情况处理
        TreeNode* left = lowestCommonAncestor(root->left, p, q); // 在左子树中查找
        TreeNode* right = lowestCommonAncestor(root->right, p, q); // 在右子树中查找
        if (left && right) return root; // 如果在两边都找到了，当前节点是LCA
        return left ? left : right; // 返回非空的子结果
    }
};

int main() {
    // 构建示例树并测试
    TreeNode *root = new TreeNode(3);
    root->left = new TreeNode(5);
    root->right = new TreeNode(1);
    root->left->left = new TreeNode(6);
    root->left->right = new TreeNode(2);
    root->right->left = new TreeNode(0);
    root->right->right = new TreeNode(8);
    root->left->right->left = new TreeNode(7);
    root->left->right->right = new TreeNode(4);

    Solution solution;
    TreeNode* p = root->left; // Node 5
    TreeNode* q = root->right; // Node 1
    TreeNode* lca = solution.lowestCommonAncestor(root, p, q);
    cout << "LCA: " << lca->val << endl; // 应输出 3，即节点3是5和1的最近公共祖先

    // 清理分配的节点
    delete root->left->right->right;
    delete root->left->right->left;
    delete root->right->right;
    delete root->right->left;
    delete root->left->right;
    delete root->left->left;
    delete root->right;
    delete root->left;
    delete root;

    return 0;
}

File: Python/2-Medium/00236._二叉树的最近公共祖先/00236._二叉树的最近公共祖先.py
from typing import Optional

class TreeNode:
    """二叉树的节点"""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def findLCA(root: Optional[TreeNode], p: TreeNode, q: TreeNode) -> Optional[TreeNode]:
    """
    寻找最近公共祖先

    Args:
    - root: TreeNode, 二叉树的根节点
    - p: TreeNode, 第一个节点
    - q: TreeNode, 第二个节点

    Returns:
    - TreeNode, 最近公共祖先节点
    """
    if root is None or root == p or root == q:
        return root
    
    left = findLCA(root.left, p, q)  # 在左子树中查找
    right = findLCA(root.right, p, q)  # 在右子树中查找

    if left and right:
        return root  # 如果p和q位于root的两侧，root即为最近公共祖先
    return left if left else right  # 返回非空的子结果

# 示例使用和测试
if __name__ == "__main__":
    # 构建示例树
    root = TreeNode(3)
    root.left = TreeNode(5)
    root.right = TreeNode(1)
    root.left.left = TreeNode(6)
    root.left.right = TreeNode(2)
    root.right.left = TreeNode(0)
    root.right.right = TreeNode(8)
    root.left.right.left = TreeNode(7)
    root.left.right.right = TreeNode(4)

    # 定位特定节点
    p = root.left  # 节点5
    q = root.right  # 节点1

    # 查找最近公共祖先
    lca = findLCA(root, p, q)
    print(f"最近公共祖先为: {lca.val}")